src/irmin-pack/unix/io_errors.ml

```diff
@@ -52,26 +99,24 @@ module Make (Io : Io.S) : S with module Io = Io = struct
   let log_error context e = [%log.err "%s failed: %a" context pp e]
 
   let catch f =
-    try Io.catch_misc_error f with _ as ex -> Base.catch (fun () -> raise ex)
+    match Base.catch (fun () -> Io.catch_misc_error f) with
+    | Ok (Ok v) -> Ok v
+    | Ok (Error e) -> Error (e :> t)
+    | Error e -> Error (e :> t)
 
   let raise_if_error = function Ok x -> x | Error e -> raise_error e
 
   let log_if_error context = function
     | Ok _ -> ()
     | Error e -> log_error context e
 
-  let io_err_result = Irmin.Type.(result int63 io_error_t)
+  let io_err_result = Irmin.Type.(result int63 T.t)
 
   let to_json_string result =
-    match result with
-    | Ok _ as v -> v |> Irmin.Type.to_json_string io_err_result
-    | Error e -> (
-        match e with
-        | `Io_misc _ as e -> Error e |> Irmin.Type.to_json_string io_err_result
-        | #error as e -> Error e |> Base.to_json_string)
+    Irmin.Type.to_json_string io_err_result (result :> (_, t) result)
 
   let of_json_string string =
     match Irmin.Type.of_json_string io_err_result string with
-    | Error (`Msg _) -> Base.of_json_string string
-    | Ok result -> (result : (_, io_error) result :> (_, [> t ]) result)
```

Comment:
This code didn't quite work to force `repr` to union the polymorphic variant. The call to `Irmin.Type.of_json_string io_err_result string` raises an "index out bounds" deep in `repr` logic (and so `Base.of_json_string` never gets a chance to run)

The fix of redefining the full polymorphic variant ain't quite satisfying, but at least the inlining is verified by the typechecker so we can't forget to update it too...